### PR TITLE
Cleanup default Functions & Tasks while uninstall+ refactor installer script

### PR DESCRIPTION
**Pull request by hossainemruz**
xref: https://github.com/stashed/stash/pull/824

### CODE REVIEW

File: deploy/stash.sh [hunk 1]
```diff
@@ -13,9 +13,61 @@ kubectl config current-context || {
...
+OS=""
+ARCH=""
+DOWNLOAD_URL=""
+DOWNLOAD_DIR=""
+TEMP_DIRS=()
+ONESSL=""
+HELM=""
+
 # http://redsymbol.net/articles/bash-exit-traps/
 function cleanup() {
```
Comment (tamalsaha) [inline deploy/stash.sh]: What if we can just add this as the installer script for the `stashed/postgres` repo and complicate the installer script for the main repo?

File: deploy/stash.sh [hunk 2]
```diff
@@ -472,6 +517,53 @@ for crd in "${non_namespaced_crds[@]}"; do
...
+  # download chart from remove repository and extract into a temporary directory
+  CHART_LOCATION="$(mktemp -dt appscode-XXXXXX)"
+  TEMP_DIRS+=(${CHART_LOCATION})
+  TEMP_INSTALLER_REPO="${CHART_NAME}-installer"
+  $HELM repo add "${TEMP_INSTALLER_REPO}" "https://charts.appscode.com/stable"
+  $HELM fetch --untar --untardir ${CHART_LOCATION} "${TEMP_INSTALLER_REPO}/${CHART_NAME}"
+  $HELM repo remove "${TEMP_INSTALLER_REPO}"
+fi
+
+$HELM template ${CHART_LOCATION}/${CHART_NAME} -x templates/update-status-function.yaml \
```
Comment (tamalsaha) [inline deploy/stash.sh]: This should be added to `https://github.com/stashed/postgres/tree/9.6/hack/deploy/install.sh`. We should leave this as the installer script for stash operator.

File: chart/stash/templates/update-status-function.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+spec:
+  image: {{ .Values.operator.registry }}/{{ .Values.operator.repository }}:{{ .Values.operator.tag }}
+  args:
+    - update-status
+    - --namespace=${NAMESPACE:=default}
+    - --repository=${REPOSITORY_NAME:=}
+    - --backup-session=${BACKUP_SESSION:=}
+    - --restore-session=${RESTORE_SESSION:=}
+    - --output-dir=${outputDir:=}
+    - --enable-status-subresource=${ENABLE_STATUS_SUBRESOURCE:=false}
```
Comment (the-redback) [inline chart/stash/templates/update-status-function.yaml]: `ENABLE_STATUS_SUBRESOURCE` is not being set to true by default.

File: deploy/stash.sh (hunk 1 above)
Comment (hossainemruz) [reply deploy/stash.sh]: removed

File: deploy/stash.sh (hunk 2 above)
Comment (hossainemruz) [reply deploy/stash.sh]: Installed this function from operator. So, no need this anymore.

File: chart/stash/templates/update-status-function.yaml (hunk 3 above)
Comment (hossainemruz) [reply chart/stash/templates/update-status-function.yaml]: This variable is resolved by Stash operator.

File: deploy/stash.sh [hunk 4]
```diff
@@ -311,6 +351,10 @@ if [ "$STASH_UNINSTALL" -eq 1 ]; then
   kubectl delete secret stash-apiserver-cert --namespace $PROMETHEUS_NAMESPACE || true
   # delete psp resources
   kubectl delete psp stash-operator-psp stash-backup-job stash-backupsession-cron stash-restore-job || true
+  # delete default functions
+  kubectl delete function update-status pvc-backup pvc-restore || true
```
Comment (tamalsaha) [inline deploy/stash.sh]: `kubectl delete function.stash.appscode.com`
`kubectl delete task.stash.appscode.com`

I think for common kind names like this, we need to use the full resource.apigroup

File: deploy/stash.sh [hunk 5]
```diff
@@ -35,37 +86,26 @@ onessl_found() {
...
-        chmod +x onessl.exe
-        export ONESSL=./onessl.exe
-        ;;
-      *)
-        echo 'other OS'
-        ;;
-    esac
-  fi
+
+  ARTIFACT="https://github.com/kubepack/onessl/releases/download/0.10.0"
```
Comment (tamalsaha) [inline deploy/stash.sh]: Why are we changing this?

Comment (hossainemruz) [reply deploy/stash.sh]: 1. If we need to change URL we won't need to change for every OS.
2. Make os detection/download part reusable. If we need to download another binary in future, we would be able to reuse .
3. Support more architecture. In future, we may need to support ARM.

File: chart/stash/templates/cleaner.yaml [hunk 6]
```diff
@@ -23,6 +23,10 @@ spec:
         command:
           - sh
           - -c
-          - "sleep 2; kubectl delete validatingwebhookconfigurations admission.stash.appscode.com || true; kubectl delete mutatingwebhookconfiguration admission.stash.appscode.com || true"
+          - "sleep 2; \
+           kubectl delete validatingwebhookconfigurations admission.stash.appscode.com || true; \
+           kubectl delete mutatingwebhookconfiguration admission.stash.appscode.com || true; \
+           kubectl delete function update-status pvc-backup pvc-restore || true; \
```
Comment (tamalsaha) [inline chart/stash/templates/cleaner.yaml]: Also update here

Comment (hossainemruz) [reply chart/stash/templates/cleaner.yaml]: fixed

**Pull request merged**